feat: add --serverless-rules to sam validate --lint

shashax42 opened the pull request:
#### Which issue(s) does this change fix?
<!-- Use the format #<issue-number>, e.g. #42 -->
This PR adds a new feature and is not associated with a specific issue number.

#### Why is this change necessary?
The AWS SAM CLI's `validate --lint` command is useful for validating CloudFormation templates, but currently lacks the ability to apply rules specific to Serverless applications. Serverless applications have different characteristics compared to general CloudFormation templates, requiring additional validation rules tailored to their specific needs.

#### How does it address the issue?
This PR adds a `--serverless-rules` option to the `sam validate --lint` command, allowing users to leverage rules from the cfn-lint-serverless package. This enables users to perform additional validations specific to Serverless applications. When enabled, this option validates Serverless application best practices such as Lambda function memory size, timeout, log retention, API Gateway stage logging, and throttling settings.

#### What side effects does this change have?
- This change does not affect existing functionality. The `--serverless-rules` option is only activated when explicitly specified.
- If the cfn-lint-serverless package is not installed, users receive a clear error message prompting them to install the package.
- When running in debug mode, additional information about Serverless Rules application and related settings is displayed.

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [x] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write/update unit tests
- [x] Write/update integration tests
- [x] Write/update functional tests if needed
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [x] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

vicheey commented on samcli/commands/validate/validate.py:
```diff
@@ -57,20 +57,27 @@ class SamTemplate:
 )
+@click.option(
+    "--serverless-rules",
+    is_flag=True,
```
Any particular region to enable this flag by default?

shashax42 replied:
@vicheey  No, there's no particular region requirement. You can enable this flag by default in all regions.

valerena commented on samcli/commands/validate/validate.py:
```diff
@@ -170,20 +199,97 @@ def _lint(ctx: Context, template: str, template_path: str) -> None:
         cfn_lint_logger.propagate = True
         cfn_lint_logger.setLevel(logging.DEBUG)
 
+    print(f"Debug info: initial linter_config = {linter_config}")
```
Don't use `print` for debugging strings, use the `logging` module instead.

```
import logging

LOG = logging.getLogger(__name__)

...

LOG.debug("Debug info: initial linter_config %s", linter_config)
```

Basic usage on this other file: https://github.com/aws/aws-sam-cli/blob/bdf9a7f19aff9064b6139c59c036cd12af01c539/samcli/lib/utils/codeuri.py#L5-L8 

https://github.com/aws/aws-sam-cli/blob/bdf9a7f19aff9064b6139c59c036cd12af01c539/samcli/lib/utils/codeuri.py#L30

Also, prefer using a message with the string format only, and the values as parameters of `debug` (like `"Debug info: initial linter_config %s", linter_config`) instead of formatting in the string directly (`f"Debug info: initial linter_config = {linter_config}"`). When `f" {parameter}"` is used for logging, the values are formatted even when logging is off, just adding extra processing that's not needed. (full details about how to use it in here: https://docs.python.org/3/library/logging.html#logging.Logger.debug)

valerena commented on samcli/commands/validate/validate.py:
```diff
@@ -170,20 +199,97 @@ def _lint(ctx: Context, template: str, template_path: str) -> None:
+    # Initialize variable to handle both options together
+    rules_to_append = []
+    
+    # Support for previous serverless_rules option (deprecated)
+    if serverless_rules:
```
Just to emphasize what I mentioned before, don't add any of the `serverless_rules` code.

valerena commented on samcli/commands/validate/validate.py:
```diff
@@ -57,20 +57,43 @@ class SamTemplate:
+    help="Specify additional lint rules to be used with cfn-lint. "
+         "Format: module.path (e.g. 'cfn_lint_serverless.rules')",
+    default=None
+)
```
I think you need to pass `multiple=True` to allow the parameter to be passed multiple times (https://click.palletsprojects.com/en/stable/options/#multiple-options and it will be stored as a list).

You mentioned in the documentation that that is actually supported, did you test passing it multiple times? (I imagine the last one will just replace any other one if we don't add the `multiple=True`)

valerena commented on samcli/commands/validate/validate.py:
```diff
@@ -170,20 +199,97 @@ def _lint(ctx: Context, template: str, template_path: str) -> None:
+        raise
 
     if not matches:
+        print(f"Debug info: template validation successful")
```
I don't think we need all these debugging messages though. Some make sense, but this part is already printing something else already.

valerena commented on tests/unit/commands/validate/test_cli.py:
```diff
@@ -1,19 +1,24 @@
 
 class TestValidateCli(TestCase):
+    def setUp(self):
+        # datetime.utcnow() 사용에 대한 경고 무시
```
Please include comments in English, thanks.

shashax42 commented:
@valerena 
## Issue: Enhance --extra-lint-rules Feature in SAM CLI (Addressing Review Feedback)

### Description
This PR enhances the `--extra-lint-rules` feature in SAM CLI, addressing all feedback from valerena.

### Addressed Feedback
1. Added `multiple=True` flag to `--extra-lint-rules` option to allow specifying multiple rule modules
2. Completely removed all code and tests related to the never-officially-released `--serverless-rules` option
3. Replaced debug `print` statements with proper `logging` module usage
4. Changed all comments to English for consistency
5. Thoroughly updated test cases to reflect the new implementation

### Test Code Improvements
- Removed obsolete `test_serverless_rules_deprecated_with_extra_lint_rules` test
- Updated all test cases to handle `extra_lint_rules` as a tuple (supporting `multiple=True`)
- Removed all references to `serverless_rules` from test parameters
- Fixed assertions to match the new parameter structure
- Changed Korean test comments to English

### Implementation Details
- Users can now use the `--extra-lint-rules` option multiple times or specify multiple modules separated by commas
- Code is cleaner and more maintainable
- Test code now correctly validates the new interface

### Related PR
#7950
